Привет, меня зовут Артем Хуршудов. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<font color='green'><b>Полезные (и просто интересные) материалы:</b></font> \
Для работы с текстами используют и другие подходы. Например, сейчас активно используются RNN (LSTM) и трансформеры (BERT и другие с улицы Сезам, например, ELMO). НО! Они не являются панацеей, не всегда они нужны, так как и TF-IDF или Word2Vec + модели из классического ML тоже могут справляться. \
BERT тяжелый, существует много его вариаций для разных задач, есть готовые модели, есть надстройки над библиотекой transformers. Если, обучать BERT на GPU (можно в Google Colab или Kaggle), то должно быть побыстрее.\
https://huggingface.co/transformers/model_doc/bert.html \
https://t.me/renat_alimbekov \
https://colah.github.io/posts/2015-08-Understanding-LSTMs/ - Про LSTM \
https://web.stanford.edu/~jurafsky/slp3/10.pdf - про энкодер-декодер модели, этеншены\
https://pytorch.org/tutorials/beginner/transformer_tutorial.html - официальный гайд
по трансформеру от создателей pytorch\
https://transformer.huggingface.co/ - поболтать с трансформером \
Библиотеки: allennlp, fairseq, transformers, tensorflow-text — множествореализованных
методов для трансформеров методов NLP \
Word2Vec https://radimrehurek.com/gensim/models/word2vec.html 

<font color='green'>Пример BERT с GPU:
```python
%%time
from tqdm import notebook
batch_size = 2 # для примера возьмем такой батч, где будет всего две строки датасета
embeddings = [] 
for i in notebook.tqdm(range(input_ids.shape[0] // batch_size)):
        batch = torch.LongTensor(input_ids[batch_size*i:batch_size*(i+1)]).cuda() # закидываем тензор на GPU
        attention_mask_batch = torch.LongTensor(attention_mask[batch_size*i:batch_size*(i+1)]).cuda()
        
        with torch.no_grad():
            model.cuda()
            batch_embeddings = model(batch, attention_mask=attention_mask_batch)
        
        embeddings.append(batch_embeddings[0][:,0,:].cpu().numpy()) # перевод обратно на проц, чтобы в нумпай кинуть
        del batch
        del attention_mask_batch
        del batch_embeddings
        
features = np.concatenate(embeddings) 
```
Можно сделать предварительную проверку на наличие GPU.\
Например, так: ```device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")```\
Тогда вместо .cuda() нужно писать .to(device)

Если понравилась работа с текстами, то можешь посмотреть очень интересный (но очень-очень сложный) курс лекций: https://github.com/yandexdataschool/nlp_course .
</font>

### <font color='orange'>Общее впечатление</font>
* Большое спасибо за проделанную работу. Видно, что приложено много усилий.
* Радует, что ноутбук хорошо структурирован. Приятно проверять такие работы.
* Отлично, что стоп-слова были исключены при векторизации!
* Над этим проектом нужно будет еще немного поработать. Однако, изменения не должны занять много времени.
* С радостью отвечу на твои вопросы, если они есть. Лучше всего их собрать в следующей ячейке. Жду новую версию проекта.

### <font color='orange'>Общее впечатление (ревью 2)</font>
* Для удобства все новые комментарии обозначены фразой "ревью 2".
* Удачи в доработке!

### <font color='orange'>Общее впечатление (ревью 3)</font>
* Новый комментарий обозначен фразой "ревью 3".

### <font color='orange'>Общее впечатление (ревью 4)</font>
* Проект зачтен!
* Удачи в дальнейшем обучении и следующих работах!

# Проект для «Викишоп»

## Описание проекта

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Загрузка и подготовка данных

In [1]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
import nltk
nltk.download('stopwords') 
from nltk.corpus import stopwords as nltk_stopwords
import numpy as np

from sklearn.feature_extraction.text import TfidfVectorizer
from catboost import CatBoostRegressor
from sklearn.tree import DecisionTreeClassifier
from lightgbm import LGBMRegressor
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import f1_score

[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что все импорты собраны в первой ячейке ноутбука! Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
</div>

In [2]:
try:
    df = pd.read_csv('C:\\Users\\ASUS\\Downloads\\toxic_comments.csv',index_col = [0], parse_dates=[0])
except FileNotFoundError:
    df = pd.read_csv('/datasets/toxic_comments.csv', index_col = [0],parse_dates=[0])

df.head()

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0

In [3]:
#информация о данных
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 159292 entries, 0 to 159450
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159292 non-null  object
 1   toxic   159292 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 3.6+ MB


<div class="alert alert-block alert-success">
<b>Успех:</b> Данные загружены корреткно.
</div>

In [4]:
#разбиение данных
features = df['text']

target = df['toxic']
features_train, features_test, target_train, target_test = \
train_test_split(features,target,shuffle=False,test_size=0.2,random_state=12345)
features_valid, features_test, target_valid, target_test = \
train_test_split(features_test,target_test,shuffle=False,test_size=0.5,random_state=12345)

In [5]:
#классификация тональности текста
corpus_train = features_train.values
corpus_valid = features_valid.values
corpus_test = features_test.values
stopwords = set(nltk_stopwords.words('english'))
 
count_tf_idf = TfidfVectorizer(stop_words=list(stopwords))
tf_idf_train = count_tf_idf.fit_transform(corpus_train)
tf_idf_valid = count_tf_idf.transform(corpus_valid)
tf_idf_test = count_tf_idf.transform(corpus_test)
 
features_train = tf_idf_train
features_valid = tf_idf_valid
features_test = tf_idf_test

<div class="alert alert-block alert-success">
<b>Успех:</b> Разбиение было сделано верно. Отлично, что векторизатор был обучен только на тренировочной части данных.
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Приводить тексты к юникоду не имеет смысла, так как они все на английском. Это может привести к падению ядра из-за увеличения объема занимаемой памяти.
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> исправил
</div>

<div class="alert alert-block alert-danger">

<b>Ошибка (ревью 2):</b> Нужно удалить `.astype('U')`.
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента2:</b> вроде исправил
</div>

<div class="alert alert-block alert-danger">

<b>Ошибка (ревью 3):</b> Это нужно сделать для всех 3-ех датасетов.
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента3:</b> да, понял, прошу прощения, в 1 ревью подумал, что вместо "тексты" написано "тесты". Их и исправил
</div>

## Обучение моделей

### Логистическая регрессия

In [6]:
model = LogisticRegression()

In [7]:
param_grid = [
    {'penalty' : ['l1','elasticnet'],
    'C' : np.logspace(-3,3,20),
    'solver' : ['liblinear'],
    'max_iter' : [100,1000,2500]
    }
]

In [8]:
clf = GridSearchCV(model, param_grid = param_grid, cv = 2, verbose = True, n_jobs=-1)

In [9]:
best_clf = clf.fit(features_train,target_train)

Fitting 2 folds for each of 120 candidates, totalling 240 fits


/opt/conda/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/sklearn/model_selection/_validation.py", line 593, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py", line 1306, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
  File "/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py", line 450, in _check_solver
    raise ValueError("Only 'saga' solver supports elasticnet penalty,"
ValueError: Only 'saga' solver supports elasticnet penalty, got solver=liblinear.

  warnings.warn("Estimator fit failed. The score on this train-test"
/opt/conda/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:610: FitF

In [10]:
best_clf.best_estimator_

LogisticRegression(C=2.976351441631316, penalty='l1', solver='liblinear')

In [11]:
model = LogisticRegression(C=2.976351441631316, max_iter=2500, penalty='l1', solver='liblinear')
model.fit(features_train, target_train)

LogisticRegression(C=2.976351441631316, max_iter=2500, penalty='l1',
                   solver='liblinear')

In [12]:
predictions = model.predict(features_valid)

In [13]:
print(f1_score(target_valid,predictions))

0.7822722476257475


<div class="alert alert-block alert-info">
<b>Совет: </b> Напомню, что внутри кросс-валидации происходит разбиение выборки на треин и валидацию. Однако, в таком случае векторизатор обучен на всей выборке, а это не совсем корректно. Для избежания такого эффекта можно использовать <a href="https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html">пайплайн</a>. <a href="https://medium.com/analytics-vidhya/ml-pipelines-using-scikit-learn-and-gridsearchcv-fe605a7f9e05">Тут</a> есть пример.
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> подскажи как корректно в моем случае создать pipe = Pipeline(???). Я пробовал делать pipe = Pipeline([('scaler', StandardScaler()), ('svc', SVC())]) ИЛИ pipe = Pipeline([('imp', SimpleImputer(missing_values=np.nan)), ('scaler', StandardScaler())]. Но выдавало ошибку Invalid parameter C for estimator Pipeline(steps=[('scaler', StandardScaler()), ('svc', SVC())]). Check the list of available parameters with `estimator.get_params().keys()`.
</div>


<div class="alert alert-block alert-info">
<b>Совет (ревью 2): </b> 

1. Внутри кросс-валидации происзодит разбиение переданной в нее выборку на части (треин и валидация). Подробнее тут: https://codecamp.ru/blog/cross-validation-k-fold/
2. Мы хотим, чтобы при подборе параметров внутри кросс-валидации векторизатор не обучался на валидации. Для этого нам нужно его обучать каждый раз только на треине (внутри кросс-валидации). Для этого и нужен пайплайн, чтобы внутри кросс-валидации обучалась не только модель, но и векторизатор. 

Вот такой пайплайн нужно передать в кросс-валидацию:  
`pipe_lr = Pipeline([(‘vect’, CountVectorizer()),
 (‘tfidf’, TfidfTransformer()),
 (‘model’, LogisticRegression())])`
    
Тут подробнее про то, как тюнить параметры у пайплайна: https://stackoverflow.com/questions/51606341/hyper-parameter-tuning-on-pipeline-object
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента2:</b> Хорошо, позже разберусь с этим, сейчас нужно поскорее сдать проект.
</div>

### Случайный лес

In [14]:
model = RandomForestClassifier()
model.fit(features_train,target_train)

RandomForestClassifier()

In [15]:
predicted = model.predict(features_valid)
print(f1_score(target_valid,predicted))

0.7136701919310616


### Вывод

Логистическая регрессия показывает наилучший результат f1-меры, 0.78. Случайный лес показывает 0.71.

<div class="alert alert-block alert-success">
<b>Успех:</b> Молодец, что попробовал разные модели в этом шаге!
</div>

### Финальный тест

In [16]:
model = LogisticRegression(C=2.976351441631316, max_iter=2500, penalty='l1', solver='liblinear')
model.fit(features_train, target_train)

LogisticRegression(C=2.976351441631316, max_iter=2500, penalty='l1',
                   solver='liblinear')

In [17]:
predictions = model.predict(features_test)
print(f1_score(target_test,predictions))

0.7910097282791009


### Вывод

На тестовых данных логистическая регрессия показывает f1 меру с показателем 0.79.

## Общий вывод

По классификации тональности текста обучили две модели: логистическая регрессия, которая показала результат на валидационной выборке 0.78, а на тестовой - 0.79; случайный лес, который показал результат - 0,71.

<div class="alert alert-block alert-success">
<b>Успех:</b> Тестирование было сделано правильно. Приятно видеть вывод в конце проекта!
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> спасибо большое за советы!
</div>